File: libexec/ssh.py
#!/usr/bin/env python
import libsite

import os, sys, getpass
import paramiko, base64
import binascii, socket, errno
import futures
from select import select
from sha import sha
from threading import Thread, Lock

from time import sleep
from random import random

from libsite.secret import vault

import signal
g_interrupted = False
def signal_handler(signum, frame):
    global g_interrupted
    g_interrupted = True
signal.signal(signal.SIGINT, signal_handler)

import logging
ll = logging.FATAL
logging.basicConfig(level=ll)
#logging.getLogger("paramiko").setLevel(ll)

#class MHKP(paramiko.MissingHostKeyPolicy):
class MHKP(paramiko.AutoAddPolicy):
    def __init__(self, *argv, **kwargs):
        paramiko.AutoAddPolicy.__init__(self, *argv, **kwargs)

    def missing_host_key(self, client, hostname, key):
        client._host_keys.add(hostname, key.get_name(), key)
        #if client._host_keys_filename is not None:
        #    client.save_host_keys(client._host_keys_filename)
        return None

class EssEssHache:
    _known_hosts = os.path.expanduser('~/.ssh/known_hosts')

    def __init__(self, threads, timeout):
        self._clients = {}
        self._threads = threads
        self._timeout = timeout

    def add_clients(self, qdns):
        for qdn in qdns:
            client = paramiko.SSHClient()
            client.load_host_keys(self._known_hosts)
            client.set_missing_host_key_policy(MHKP())
            self._clients[qdn] = client

    def del_client(self, qdn):
        client = self._clients.pop(qdn)
        if client._host_keys_filename is not None:
            client.save_host_keys(client._host_keys_filename)
        client.close()

    def connect(self, username, qdn, command, stdin=[]):
        #sleep(2 * random())

        client = self._clients[qdn]
        client.known_hosts = None
        data = None
        try:
            data = (qdn, [], [], 1)
            try:
                ssh_proxy_host = os.environ.get('SSH_PROXY_HOST', None)
                ssh_proxy_port = os.environ.get('SSH_PROXY_PORT', '22')
                tld = os.environ.get('TLD', None)
                #sys.stdout.write("%s=%s\n" % ('ssh_proxy_host', ssh_proxy_host))
                #sys.stdout.write("%s=%s\n" % ('tld', tld))

                if not ssh_proxy_host:
                    client.connect(qdn, username=username, timeout=self._timeout)
                else:
                    #sys.stderr.write("#. >>> Queued up %s via proxy %s\n" % (qdn, ssh_proxy_host))
                    proxy_command = paramiko.ProxyCommand(
                        "ssh -p %s %s@%s nc %s.%s 22" % (
                            ssh_proxy_port, username, ssh_proxy_host, qdn, tld
                        )
                    )
                    client.connect(qdn, username=username, sock=proxy_command, timeout=self._timeout)

                stdout = []
                stderr = []
                ecode = -1

                if stdin:
                    chan = client.get_transport().open_session()
                    _stdin = chan.makefile('wb')
                    _stdout = chan.makefile('rb')
                    _stderr = chan.makefile_stderr('rb')

                    #. If stdout is still open then sudo is asking us for a password
                    if True not in [_stdin.channel.closed, _stdout.channel.closed]:
                        chan.exec_command(command)

                        for line in stdin:
                            _stdin.write("%s\n" % line)
                        _stdin.flush()

                        ecode = chan.recv_exit_status()
                        stdout = _stdout.readlines()
                        stderr = _stderr.readlines()
                    else:
                        estr = "StdinClosedButData"
                        stderr = [estr]
                        stdout = []
                        ecode = 4
                else:
                    _stdin, _stdout, _stderr = client.exec_command(command)
                    ecode = _stdout.channel.recv_exit_status()
                    stdout = _stdout.readlines()
                    stderr = _stderr.readlines()

                data = (qdn, stdout, stderr, ecode)

            except paramiko.BadHostKeyException as exception:
                estr = "Bad Host Key"
                data = (qdn, [], [estr], 2)

            except paramiko.SSHException as exception:
                estr = "Unknown Host"
                data = (qdn, [], ["%s: %s" % (estr, str(exception))], 2)

            except socket.timeout as exception:
                estr = "TCP Socket Timeout"
                data = (qdn, [], [estr], 3)

            except socket.error as exception:
                if exception.errno == errno.EHOSTUNREACH:
                    estr = "Host Unreacheable"
                    data = (qdn, [], [estr], 4)
                elif exception.errno == errno.ETIMEDOUT:
                    estr = "Timeout"
                    data = (qdn, [], [estr], 4)
                elif exception.errno == errno.ECONNRESET:
                    estr = "Connection Reset"
                    data = (qdn, [], [estr], 4)
                else:
                    estr = "C %s: %s (%s)" % (exception.__class__, exception, exception.errno)
                    data = (qdn, [], [estr], 7)

            except Exception as exception:
                estr = "B %s: %s" % (exception.__class__, exception)
                data = (qdn, [], [estr], 8)

            else:
                self.del_client(qdn)

        except Exception as exception:
            estr = "A %s: %s" % (exception.__class__, exception)
            data = (qdn, [], [estr], 9)

        return data


    def writeMeABASHScript(self, var, username, password, cmd):
        mutex = Lock()

        sys.stdout.write("#!/bin/bash\n")
        sys.stdout.write("#. threads: %d\n" % self._threads)
        sys.stdout.write("#. timeout: %0.1f\n" % self._timeout)
        sys.stdout.write("\n")
        with futures.ThreadPoolExecutor(max_workers=self._threads) as executor:
            #. See if stdin from the shell has any data to offer...
            #stdindata = []
            #if select([sys.stdin,],[],[],0.0)[0]:
            #    stdindata = [_.strip('\n') for _ in sys.stdin.readlines()]

            stdindata = []
            if password:
                cmd = """sudo -S %s""" % cmd
                stdindata.append(password)

            queue = {
                executor.submit(self.connect, username, qdn, cmd, stdindata):
                    qdn for qdn in self._clients.keys()
            }

            sys.stdout.write("local -A %s\n"   % var) #. stdout
            sys.stdout.write("local -A %s_o\n" % var) #. stdout
            sys.stdout.write("local -A %s_e\n" % var) #. stderr
            sys.stdout.write("local -A %s_w\n" % var) #. warnings

            i = 0
            total = len(self._clients.keys())

            #for future in futures.as_completed(queue, timeout=self._timeout):
            for future in futures.as_completed(queue):
                i += 1

                qdn = queue[future]
                data = None
                try:
                    #data = future.result(timeout=self._timeout)
                    data = future.result()
                except futures.TimeoutError:
                    sys.stdout.write("#. %s is taking too long. Oh well.\n" % qdn)
                else:
                    mutex.acquire()
                    try:
                        sys.stdout.write("#. %s -={\n" % qdn)

                        _qdn, stdout, stderr, e = data
                        qdn_hash = sha(qdn).hexdigest()
                        sys.stdout.write(
                            "#. Query with host %s (%d of %d) : %s\n" % (
                                qdn, i, total, future._state
                            )
                        )

                        sys.stdout.write("%s[%s]=%d\n" % (var, qdn_hash, e))

                        if len(stdout) > 0:
                            sys.stdout.write("read -r -d '' %s_o[%s] <<-!\n" % (var, qdn_hash))
                            sys.stdout.write(''.join(stdout).strip())
                            sys.stdout.write("\n!\n")

                        if len(stderr) > 0:
                            sys.stdout.write("read -r -d '' %s_e[%s] <<-!\n" % (var, qdn_hash))
                            sys.stdout.write(''.join(stderr).strip())
                            sys.stdout.write("\n!\n")

                        sys.stdout.write("#. }=- %s\n\n" % qdn)
                    finally:
                        mutex.release()

            sys.stdout.write("#. All done.\n")

def main():
    '''
    Assumptions:

        1. You have ssh public keys in place to connect to remote hosts
        2. If you supply a sudo secret id, you want to wrap the command in sudo
    '''
    e=1

    usage = """
Usage: ssh <threads> <timeout> <var>=<username>[:<sudo-secret-id>]@<host>,<host>,<host>,... <cmd>
"""

    if len(sys.argv) >= 3:
        threads = int(sys.argv[1])
        timeout = float(sys.argv[2])

        var = None
        username = None
        qdns = list()
        cmd = None

        var, remainder = sys.argv[3].split('=', 1)
        password = None
        username, remainder = remainder.split('@', 1)
        if ':' in username:
            username, sid = username.split(':', 1)
            password = secret(sid)

        qdns = remainder.split(',')
        cmd = ' '.join(sys.argv[4:])

        ssh = EssEssHache(threads, timeout)
        ssh.add_clients(qdns)

        ssh.writeMeABASHScript(var, username, password, cmd)

        e=0
    else:
        print(usage)

    return e

if __name__ == '__main__':
    e = 1
    try:
        e = main()
    except KeyboardInterrupt:
        e = 130
    except (IOError, OSError):
        e = 141

    sys.exit(e)

File: libexec/ssh.rb
#!/usr/bin/env ruby
require 'resolv'
require 'digest/sha1'

require 'net/ssh'
require 'net/ssh/multi'
require 'net/ssh/gateway'

require 'parallel'
#. }=-

class EssEssHache
  @@dns = Resolv::DNS.open

  def initialize(threads, attempts, timeout)
    @threads  = threads
    @attempts = attempts
    @timeout  = timeout
    @clients  = []
    @sessions = {}
    @gateway  = nil
    @username = nil
    @context  = nil
    @command  = "hostname"
    @sudo     = false
    @sudosec  = nil
    @tries    = 0
  end

  def add_clients(clients)
    @clients += clients
  end

  def set_username(username)
    @username = username
  end

  def set_context(context)
    @context = context
  end

  def set_proxy(host, port)
    @gateway = Net::SSH::Gateway.new(
      host, @username, port: port, :verbose => :fatal,
    )
    return @gateway.active?
  end

  def set_command(cmd)
    @command = cmd
  end

  def set_sudo(duso, secret=nil)
    @sudo = duso
    @sudosec = secret
  end

  def connect(host)
    @sessions[host] = {
      retcode: -1,  stdout: '',
      sigcode: -1,  stderr: '',
      session: nil, except: '',
      success: nil, status: '',
    }

    begin
      if @gateway
        #@sessions[host][:session] = Net::SSH::Gateway.new(
        #  'night.sirca.org.au',
        #  @username,
        #  :port => 222,
        #  :compression => "none",
        #  :auth_methods => [ 'publickey' ],
        #  :verbose => :fatal,
        #).ssh(host, @username,
        #  :timeout => @timeout,
        #  :compression => "none",
        #  :auth_methods => [ 'publickey' ],
        #  :verbose => :fatal,
        #)
        @sessions[host][:status] = "Connecting to #{host} via the gateway #{@gateway}..."
        @sessions[host][:session] = @gateway.ssh(host, @username,
          :timeout => @timeout,
          :compression => "none",
          :auth_methods => [ 'publickey' ],
          :verbose => :fatal,
        )
      else
        @sessions[host][:status] = 'Connecting...'
        #@@dns.getaddress host
        @sessions[host][:session] = Net::SSH.start(host, @username,
          :config => [
            "#{ENV['HOME']}/.site/etc/ssh.conf",
            "#{ENV['HOME']}/.ssh/config",
            "/etc/ssh_config"
          ],
          :timeout => @timeout,
          :compression => "none",
          :auth_methods => [ 'publickey' ],
          :verbose => :fatal,
        )
      end

      rescue Resolv::ResolvError => ouch
        @sessions[host][:except] = 'DNS Error'
        @sessions[host][:status] += 'DNS Error'
      rescue Timeout::Error
        @sessions[host][:except] = "Connection Timeout"
        @sessions[host][:status] += 'Connection Failed'
      rescue Errno::EHOSTUNREACH
        @sessions[host][:except] = "Host unreachable"
        @sessions[host][:status] += 'Connection Failed'
      rescue Errno::ECONNREFUSED
        @sessions[host][:except] = "Connection Refused"
        @sessions[host][:status] += 'Connection Failed'
      rescue Net::SSH::AuthenticationFailed
        @sessions[host][:except] = "Authentication Failure"
        @sessions[host][:status] += 'Connection Failed'
      rescue Net::SSH::Disconnect
        @sessions[host][:except] = "SSH Disconnect"
        @sessions[host][:status] += 'Connection Failed'
      rescue
        if @tries < @attempts
          @tries += 1
        else
          retry
        end
      else
        @sessions[host][:status] += 'Connected'
    end
  end

  def exec(host, command)
    command = "sudo -S #{command}" if @sudo
    #. Submit the job, don't wait for it however:

    if not @sessions[host][:session].nil?
      @sessions[host][:success] = false
      begin
        @sessions[host][:status] += "; Executing..."
        @sessions[host][:session].open_channel do |channel|
          @sessions[host][:status] = 'Executing (requesting pty)...'
          channel.request_pty do |ch0, success|
            if success
              @sessions[host][:status] = 'Executing command...'
              channel.exec(command) do |ch1, success|
                if success
                  @sessions[host][:status] = 'Execution submitted'
                  channel.on_data do |ch3, data|
                    if @sudo and @sudosec and data =~ /^\[sudo\] password for/
                      channel.send_data "#{@sudosec}\n"
                    else
                      @sessions[host][:stdout] += data
                    end
                  end
                  channel.on_open_failed do |ch, code, desc|
                    p code, desc
                  end
                  channel.on_extended_data do |ch, data|
                    @sessions[host][:stderr] += "#{data.inspect}"
                  end
                  channel.on_request('exit-status') do |ch, data|
                    @sessions[host][:retcode] = data.read_long
                  end
                  channel.on_request('exit-signal') do |ch, data|
                    @sessions[host][:sigcode] = data.read_long
                  end
                  channel.on_close do |ch|
                    @sessions[host][:success] = true
                  @sessions[host][:status] = 'Execution complete'
                  end
                else
                  @sessions[host][:status] = 'Execution failed'
                  @sessions[host][:except] = "Execution Failure"
                  @sessions[host][:retcode] = -2
                end
              end
            else
              @sessions[host][:status] = 'Execution failed to obtain PTY'
              @sessions[host][:except] = "Failure to obtain PTY"
              @sessions[host][:retcode] = -3
            end
          end
        end
      rescue
        @sessions[host][:status] = 'Execution failed with unknown exception'
        @sessions[host][:except] = "Unhandled exception"
        @sessions[host][:retcode] = -9
      end
    else
      @sessions[host][:status] += '; No session, no execution'
    end
  end

  def harvest
    printf("#. Establishing connectivity to %d hosts...", @clients.length)
    Parallel.map(@clients, :in_threads=>@threads) { |host| self.connect(host) }
    printf("DONE\n")

    printf("#. Submitting commands to %d hosts...", @sessions.length)
    @sessions.map { |host,hostdata| self.exec(host, @command) }
    #Parallel.map(@sessions, :in_threads=>@threads) { |host,hostdata| self.exec(host, @command) }
    printf("DONE\n")

    condition = Proc.new { |ssh| ssh.busy? }
    wait = true
    while wait
      wait = false
      #printf "#. Waiting on %d connections...\n", @sessions.length
      #@sessions.each do |host,hostdata|
      #  printf "#{host} #{hostdata[:status]} #{@gateway.zits.forward.active_locals}\n"
      #  #pp @sessions[host].reject { |k,v| k == :session }
      #end
      @sessions.each do |host,hostdata|
        if hostdata[:session] and hostdata[:session].process(1, &condition)
          wait = true
          break
        end
      end
    end
  end

  def close
    @sessions.each do |host,hostdata|
      hostdata[:session].close unless hostdata[:session].nil?
    end
    @gateway.shutdown! unless @gateway.nil?
  end

  def writeMeABASHScript
    printf("#!/bin/bash\n")
    printf("#. %d threads\n", @threads)
    printf("#. %d attempts\n", @attempts)
    printf("#. %0.1fs timeout\n", @timeout)
    printf("\n")

    printf "local -A %s\n", @context
    printf "local -A %s_o\n", @context
    printf "local -A %s_e\n", @context
    printf "local -A %s_w\n", @context

    @sessions.each do |host, hostdata|
      printf "#. %s...\n", host

      hash = Digest::SHA1.hexdigest host
      printf("%s[%s]=%d\n", @context, hash, hostdata[:retcode])

      stdout = hostdata[:stdout].strip
      stderr = hostdata[:stderr].strip
      except = hostdata[:except].strip

      printf(
        "read -r -d '' %s_o[%s] <<-!\n%s\n!\n",
        @context, hash, stdout
      ) if stdout.length > 0

      if stderr.length > 0
        printf(
          "read -r -d '' %s_e[%s] <<-!\n%s\n!\n",
          @context, hash, stderr
        )
      elsif except.length > 0
        printf(
          "read -r -d '' %s_e[%s] <<-!\nException: %s\n!\n",
          @context, hash, except
        )
      end
    end
  end
end

def secret(sid)
  require 'gpgme'

  encrypted_data = GPGME::Data.new(
    File.open("#{ENV['HOME']}/.site/etc/site.vault")
  )
  ctx = GPGME::Ctx.new
  decrypted = ctx.decrypt encrypted_data
  decrypted.seek(0)

  secrets = Hash[
    *decrypted.read.split("\n").collect { |v| v.split(/\s+/, 2) }.flatten
  ]

  return secrets[sid]
end

def main
  if ARGV.length >= 4
    threads = ARGV[0].to_i
    attempts = ARGV[1].to_i
    timeout = ARGV[2].to_f

    ssh = EssEssHache.new(threads, attempts, timeout)

    context, remainder = ARGV[3].split('=', 2)
    password = nil

    if ENV.key? 'SUDO' and ENV['SUDO'].length > 0
      password = secret(ENV['SUDO'])
      ssh.set_sudo(true, password)
    end

    clients = remainder.split(',').uniq
    printf("#. %d clients!\n", clients.length)
    cmd = ARGV[4..-1].join(' ')

    ssh.set_command(cmd)
    ssh.set_context(context)

    #ssh_proxy_host = ENV['SSH_PROXY_HOST']
    #if ssh_proxy_host.length > 0
    #  ssh_proxy_port = ENV['SSH_PROXY_PORT'].to_i or 22
    #  ssh.set_proxy(ssh_proxy_host, ssh_proxy_port)
    #end

    ssh.add_clients(clients)
    ssh.harvest
    ssh.writeMeABASHScript
    ssh.close
  end
end

main
